contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -1199,6 +1206,135 @@ object_id(PG_FUNCTION_ARGS)
+		if(db_name == NULL) /* database doesn't exist with given oid */
+			PG_RETURN_NULL();
+		user_id = GetSessionUserId();
+	}
+
+	/* search in list of ENRs registered in the current query environment by object_id */
+	enr = get_ENR_withoid(currentQueryEnv, object_id);
+	if(enr != NULL && enr->md.enrtype == ENR_TSQL_TEMP)
+	{
+		result = pstrdup(enr->md.name);
+		PG_RETURN_VARCHAR_P((VarChar *) cstring_to_text(result));
+	}
+
+	/* search in pg_class by object_id */
+	tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(object_id));
+	if (HeapTupleIsValid(tuple))
+	{
+		/* check if user have right permission on object */
+		if (pg_class_aclcheck(object_id, user_id, ACL_SELECT) == ACLCHECK_OK)
+		{	
+			Form_pg_class pg_class = (Form_pg_class) GETSTRUCT(tuple);
+			result = pstrdup(NameStr(pg_class->relname));
+			schema_id = pg_class->relnamespace;
+		}
+		ReleaseSysCache(tuple);
+		found = true;
+	}
+
+	if (!found)
+	{
+		/* search in pg_proc by object_id */
+		tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(object_id));
+		if (HeapTupleIsValid(tuple))
+		{
+			/* check if user have right permission on object */
+			if (pg_proc_aclcheck(object_id, user_id, ACL_EXECUTE) == ACLCHECK_OK)
+			{
+				Form_pg_proc procform = (Form_pg_proc) GETSTRUCT(tuple);
+				result = pstrdup(NameStr(procform->proname));
+				schema_id = procform->pronamespace;
+			}
+			ReleaseSysCache(tuple);
+			found = true;
+		}
+	}
+
+	if (!found)
+	{
+		/* search in pg_types by object_id */
+		tuple = SearchSysCache1(TYPEOID, ObjectIdGetDatum(object_id));
+		if (HeapTupleIsValid(tuple))
+		{
+			/* check if user have right permission on object */
+			if (pg_type_aclcheck(object_id, user_id, ACL_USAGE) == ACLCHECK_OK)
+			{	
+				Form_pg_type pg_type = (Form_pg_type) GETSTRUCT(tuple);
+				result = pstrdup(NameStr(pg_type->typname));
+			}
+			ReleaseSysCache(tuple);
+			found = true;
+		}
+	}
+	
+	if(!found)
+	{
+		/* search in pg_trigger by object_id */
+		tgrel = table_open(TriggerRelationId, AccessShareLock);
+		ScanKeyInit(&key,
+				Anum_pg_trigger_oid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(object_id));
+
+		tgscan = systable_beginscan(tgrel, TriggerOidIndexId, true,
+						NULL, 1, &key);
+
+		tuple = systable_getnext(tgscan);
+		if (HeapTupleIsValid(tuple))
+		{
+			Form_pg_trigger pg_trigger = (Form_pg_trigger) GETSTRUCT(tuple);
+			/* check if user have right permission on object */
+			if(OidIsValid(pg_trigger->tgrelid) && 
+				pg_class_aclcheck(pg_trigger->tgrelid, user_id, ACL_SELECT) == ACLCHECK_OK)
+			{
+				result = pstrdup(NameStr(pg_trigger->tgname));
+				schema_id = get_rel_namespace(pg_trigger->tgrelid);
+			}
+			found = true;
+		}
+		systable_endscan(tgscan);
+		table_close(tgrel, AccessShareLock);
+	}
+
+	if(result)
+	{	
+		/* check if schema corresponding to found object belongs to specified database */
+		if(!OidIsValid(schema_id) || is_schema_from_db(schema_id, database_id)) /* in case of pg_type schema_id will be invalid */
```

Comment:
is there a specific test case to cover this? schema is part of specified db vs it is not

Reply:
yes, cross-db lookup test consider those scenario